File: p2_procesamiento_por_lotes_multiprogramacion/src/kb.rs
use crate::screen;

use std::io::{self, Read,Write};
use std::{thread, time}; // For system sleep
use termion::event::Key;
use termion::input::TermRead;
use termion::raw::IntoRawMode; // Import IntoRawMode trait
use termion::terminal_size;

pub fn bind() {
    let stdin = io::stdin();
    let mut stdout = io::stdout().into_raw_mode().unwrap();
    let mut keys = stdin.keys();

    let mut paused = false;

    loop {
        if let Some(Ok(key)) = keys.next() {
            stdout.flush().unwrap();

            if paused {
                if let Key::Char('c') = key {
                    print_status("Continue");
                } else {
                    continue;
                }
            } else {
                match key {
                    Key::Char('p') => {
                        print_status("Paused");
                        paused = true;
                        continue;
                    }
                    Key::Char('w') => {
                        print_status("Error");
                    }
                    Key::Char('e') => {
                        print_status("Interruption");
                    }
                    _ => { continue; }
                }
            }

            break;
        }

    }
    stdout.flush().unwrap();
}

fn print_status(text: &str) {
    let (term_width, _term_height) = terminal_size().unwrap();
    let x_top_right = term_width - text.len() as u16 - 1;
    screen::print_xy(x_top_right, 1, text);

    if text != "Paused" {
        let spaces: String = " ".repeat(text.len());
        thread::sleep(time::Duration::from_secs(1));
        screen::print_xy(x_top_right, 1, &spaces);
    }
}

File: p2_procesamiento_por_lotes_multiprogramacion/kb.py
import select
import sys
import tty
import termios
import time

# My files
import screen


def kbhit(time_target, previous_time):
    def is_key_pressed():
        return select.select([sys.stdin], [], [], 0) == ([sys.stdin], [], [])

    # Set terminal to raw mode
    old_settings = termios.tcgetattr(sys.stdin)
    try:
        tty.setcbreak(sys.stdin.fileno())

        key = 0
        start = time.time()
        duration = 0

        while key == 0 and duration < time_target:
            duration = time.time() - start + previous_time
            if is_key_pressed():
                key = sys.stdin.read(1)

        # print(f"{duration}")

    finally:
        # Restore terminal settings
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_settings)

    return key, duration


def print_status(text):
    screen.print_top_right(text)

    if text != "Paused":
        spaces = " " * len(text)
        time.sleep(1)
        screen.print_top_right(spaces)


def bind(time_target):
    key = 0
    duration = 0
    previous_time = 0

    while duration < time_target:
        key, duration = kbhit(time_target, previous_time)

        if key == "p":
            print_status("Paused")
            previous_time = duration

            while True:
                key, _ = kbhit(time_target, previous_time)
                if key == "c":
                    print_status("Continue")
                    break

            continue
        elif key == "w":
            print_status("Error")
        elif key == "e":
            print_status("Interruption")
        else:
            continue

        break

    sys.stdout.flush()
    return key, int(duration)
